build.sbt

```diff
@@ -2,198 +2,221 @@ import sbt.Keys._
 
 import java.io.File
 
-def doServiceSetup(project: Project, folder: String) =
-  project
-    .in(new File(folder))
-    .dependsOn(common % "compile->compile;test->test")
-    .settings(Common.settings: _*)
-    .settings(Finatra.settings: _*)
-    .settings(DockerCompose.settings: _*)
-    .enablePlugins(DockerComposePlugin)
-    .enablePlugins(JavaAppPackaging)
+def setupProject(
+  project: Project,
+  folder: String,
+  localDependencies: Seq[Project] = Seq(),
+  externalDependencies: Seq[ModuleID] = Seq()
+): Project = {
+
+  // Here we write a bit of metadata about the project, and the other
+  // local projects it depends on.  This can be picked up by some hypothetical
+  // future build system to determine whether to run tests based on the
+  // up-to-date project graph.
+  // See https://www.scala-sbt.org/release/docs/Howto-Generating-Files.html
+  val file = new File(s"builds/sbt_metadata/${project.id}.json")
+  val dependenciesJson = localDependencies
+    .map { p: Project => p.id }
+    .map { id: String => s""""$id""""}
+    .mkString(", ")
+
+  IO.write(file, s"""
```

Comment:
Why are we writing JSON like this? You've got the whole Scala ecosystem to play with.

Reply:
Running inside sbt seems to be missing a bunch of bits – in particular, I couldn’t find `FileWriter` or `JSONObject`.

Maybe they need importing if you don't have the standard library by default? It’s only just occurred to me to try – I was trying to get something done quickly.

Reply:
Also I think there are pre-built SBT plugins which do this: https://github.com/sbt/sbt-buildinfo

Reply:
Ooh, much nicer

Reply:
And now you ask, I’ve put in a proper JSON encoder.